varnishreload

```diff
@@ -0,0 +1,216 @@
+# are met:
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+
+set -e
+set -u
+
+MAXIMUM=
+VCL_FILE=
+VCL_LABEL=
+VCL_NAME=
+WARMUP=
+WORK_DIR=
+
+SCRIPT="$0"
+
+usage() {
+	test $# -eq 1 &&
+	printf 'Error: %s.\n\n' "$1"
+
+	cat <<-EOF
+	Usage: $SCRIPT [-l <label>] [-m <max>] [-n <workdir>] [-w <warmup>] [<file>]
+	       $SCRIPT -h
+
+	Reload and use a VCL on a running Varnish instance.
+
+	Available options:
+	-h           : show this help and exit
+	-l <label>   : name of the VCL label to reload
+	-m <max>     : maximum number of available reloads to leave behind
+	-n <workdir> : specify the name or directory for the varnishd instance
+	-w <warmup>  : the number of seconds between load and use operations
+
+	When <label> is empty or missing, the active VCL is reloaded. If the
+	reloaded VCL is a label, the underlying VCL program is reloaded, and
+	the label is updated to reference the new program.
+
+	When <file> is empty or missing, the selected VCL's file is used but
+	reload will fail if it wasn't loaded from a file. The <file>, when
+	specified, is passed as-is to the vcl.load command. Refer to the
+	varnishd manual regarding the handling of absolute or relative paths.
+
+	Upon success, the name of the loaded VCL is constructed from the
+	current date and time, for example:
+
+	    $(vcl_reload_name)
+
+	Afterwards available VCLs created by this script are discarded until
+	<max> are left, unless it was empty or undefined. VCLs referenced by
+	a label aren't discarded.
+	EOF
+	exit $#
+}
+
+varnishadm() {
+	if ! OUTPUT=$(command varnishadm -n "$WORK_DIR" -- "$@" 2>&1)
+	then
+		echo "Command: varnishadm -n '$WORK_DIR' -- $*"
+		echo
+		echo "$OUTPUT"
+		echo
+		return 1
+	fi >&2
+	echo "$OUTPUT"
+}
+
+fail() {
+	echo "Error: $*" >&2
+	exit 1
+}
+
+vcl_reload_name() {
+	printf "reload_%s" "$(date +%Y%m%d_%H%M%S)"
```

Comment:
You might want a _%N at the end of the format string, otherwise you can't reload twice during the same second.  (My autopkgtest suite does that)

Reply:
My stance on it was clearly nope with the first version of the script (see https://github.com/varnishcache/pkg-varnish-cache/issues/76#issuecomment-306676097 for example) but as I was about to reply the same to you, it occurred to me that reloading several VCLs in the same second now makes sense if you want to reload several labels at once.

Thanks!

Reply:
Oh great, `%N` is not portable, looking for an alternative.

Reply:
For reference, I first checked [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/date.html) and upon not seeing `%N` looked at [FreeBSD](http://www.unix.com/man-page/FreeBSD/1/date/) to see whether I could close my eyes on this detail, but I can't...